### PR TITLE
[TE] anomaly detector interface change and implementation for rule-based detection

### PR DESCRIPTION
**Pull request by jihaozh**
- Update the anomaly detector interface to return a detection result which includes anomalies and time series (containing time stamps, predicted baseline, current value, upper/lower bounds)
- Implement the new detector interface in Threshold, Percentage Change, Absolute Change, and Holt-Winters detector.
- Implement the baseline provider interfaces in Threshold, Percentage Change, Absolute Change classes.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/DetectionResult.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package org.apache.pinot.thirdeye.detection.spi.model;
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/DetectionResult.java]: Add headers.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/DetectionResult.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import org.apache.pinot.thirdeye.datalayer.dto.MergedAnomalyResultDTO;
+
+
+/**
+ * The detection result
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/DetectionResult.java]: Better to add more description.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java [hunk 3]
```diff
@@ -105,11 +109,44 @@
 
     // make anomalies
     DatasetConfigDTO datasetConfig = data.getDatasetForMetricId().get(me.getId());
-    return DetectionUtils.makeAnomalies(slice, df, COL_ANOMALY, window.getEndMillis(),
+    List<MergedAnomalyResultDTO> anomalies = DetectionUtils.makeAnomalies(slice, df, COL_ANOMALY, window.getEndMillis(),
         DetectionUtils.getMonitoringGranularityPeriod(monitoringGranularity, datasetConfig), datasetConfig);
+    return DetectionResult.from(anomalies, TimeSeries.fromDataFrame(constructAbsoluteChangeBoundaries(df)));
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java]: I would suggest to put baseline calculation in a separate line so it is easier to debug.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java [hunk 4]
```diff
@@ -105,11 +109,44 @@
...
 
   @Override
+  public TimeSeries computePredictedTimeSeries(MetricSlice slice) {
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(this.baseline.scatter(slice)));
+    DataFrame dfBase = this.baseline.gather(slice, data.getTimeseries());
+    return TimeSeries.fromDataFrame(constructAbsoluteChangeBoundaries(dfBase));
+  }
+
+  private DataFrame constructAbsoluteChangeBoundaries(DataFrame dfBase) {
+    if (!Double.isNaN(this.absoluteChange)) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java]: Why we have the check here? When will absoluteChange would be NaN?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java [hunk 5]
```diff
@@ -87,8 +91,38 @@
...
+  @Override
+  public TimeSeries computePredictedTimeSeries(MetricSlice slice) {
+    InputData data =
+        this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(Collections.singletonList(slice)));
+    DataFrame df = data.getTimeseries().get(slice);
+    return TimeSeries.fromDataFrame(constructThresholdBoundaries(df));
+  }
+
+  private DataFrame constructThresholdBoundaries(DataFrame df) {
+    if (!Double.isNaN(this.min)) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: When will this.min be NaN?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java [hunk 6]
```diff
@@ -87,8 +91,38 @@
...
+      df.addSeries(COL_LOWER_BOUND, DoubleSeries.fillValues(df.size(), this.min));
+    }
+    if (!Double.isNaN(this.max)) {
+      df.addSeries(COL_UPPER_BOUND, DoubleSeries.fillValues(df.size(), this.max));
+    }
+    df.addSeries(COL_VALUE, DoubleSeries.fillValues(df.size(), predictedValue()));
+    return df;
+  }
+
+  private double predictedValue() {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: There should be no predicted value (or the predicted value is the current value) for threshold based detector.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/TimeSeries.java [hunk 7]
```diff
@@ -62,17 +71,34 @@ private static void addSeries(TimeSeries ts, DataFrame df, String name) {
...
+    return ts;
+  }
+
   /**
    * Add DataFrame into TimeSeries.
    * @param df The source DataFrame.
    * @return TimeSeries that contains the predicted values.
    */
   public static TimeSeries fromDataFrame(DataFrame df) {
+    Preconditions.checkArgument(df.contains(COL_TIME));
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/spi/model/TimeSeries.java]: Good job to have this precondition check!

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/wrapper/AnomalyDetectorWrapper.java [hunk 8]
```diff
@@ -187,7 +187,7 @@ public DetectionPipelineResult run() throws Exception {
         LOG.warn("[DetectionConfigID{}] detecting anomalies for window {} to {} failed.", this.config.getId(), window.getStart(), window.getEnd(), e);
         lastException = e;
       }
-      anomalies.addAll(anomaliesForOneWindow);
+      anomalies.addAll(detectionResult.getAnomalies());
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/wrapper/AnomalyDetectorWrapper.java]: I am confused here only the anomalies are merged.
What about the baselines in detectionResult?

**Comment by xiaohui-sun** [comment]
Thanks for adding this. Seems this is not a complete solution. Do you plan to add more PRs later to expose DetectionResult out?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java (hunk 4 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java]: By default, it's NaN which means not set. If it's NaN, no anomaly will be generated.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java (hunk 5 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: By default, it's NaN which means not set. It's possible only one side threshold is set

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java (hunk 6 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: Good point. will change to the same value as the current.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/wrapper/AnomalyDetectorWrapper.java (hunk 8 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/wrapper/AnomalyDetectorWrapper.java]: Those will be handled in a follow-up PR.

**Comment by jihaozh** [comment]
> Thanks for adding this. Seems this is not a complete solution. Do you plan to add more PRs later to expose DetectionResult out?

Yes. This PR focus on the interface change and implementation for the rule-based detectors.
I will submit follow up PRs to do the following:
- Implement the interface in ALGORITHM detectors.
- Change the detection pipeline to handle the baseline time series.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java (hunk 4 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/AbsoluteChangeRuleDetector.java]: I see. It is a little weird to use NaN as default value.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java (hunk 6 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: One thing is that, if we set the predicted value the same as current value, the percentage change will be 0 in UI.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java (hunk 5 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: Got it. Thanks.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java (hunk 6 above)
Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: Why we show percentage change for threshold based rules?

Comment (xiaohui-sun) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/ThresholdRuleDetector.java]: Let's discuss that offline and maybe submit another PR to fix this. @jihaozh

**Pull request merged**